File: docs/examples/webhook-verify/verify.js
#!/usr/bin/env node
"use strict";

/* eslint-disable @typescript-eslint/no-require-imports */
const crypto = require("node:crypto");
const http = require("node:http");

const secret = process.env.CONVO_WEBHOOK_SECRET;
const port = Number(process.env.PORT || 3000);
const toleranceSeconds = 300;

if (!secret) {
  console.error("Set CONVO_WEBHOOK_SECRET before starting the server.");
  process.exit(1);
}

function parseSignatureHeader(header) {
  if (!header) return null;

  const fields = new Map();
  for (const part of header.split(",")) {
    const pieces = part.split("=");
    if (pieces.length !== 2) return null;
    fields.set(pieces[0], pieces[1]);
  }

  const timestampRaw = fields.get("t");
  const signature = fields.get("v1");
  if (!timestampRaw || !/^\d+$/.test(timestampRaw)) return null;
  if (!signature || !/^[a-f0-9]{64}$/.test(signature)) return null;

  return {
    timestamp: Number(timestampRaw),
    signature,
  };
}

function signBody(timestamp, body) {
  return crypto
    .createHmac("sha256", secret)
    .update(`${timestamp}.${body}`, "utf8")
    .digest("hex");
}

function timingSafeEqualHex(left, right) {
  const leftBuffer = Buffer.from(left, "hex");
  const rightBuffer = Buffer.from(right, "hex");
  if (leftBuffer.length !== rightBuffer.length) return false;
  return crypto.timingSafeEqual(leftBuffer, rightBuffer);
}

function verifySignature(header, body) {
  const parsed = parseSignatureHeader(header);
  if (!parsed || !Number.isSafeInteger(parsed.timestamp)) return false;

  const now = Math.floor(Date.now() / 1000);
  if (Math.abs(now - parsed.timestamp) > toleranceSeconds) return false;

  const expected = signBody(parsed.timestamp, body);
  return timingSafeEqualHex(parsed.signature, expected);
}

function readBody(request) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    request.on("data", (chunk) => chunks.push(chunk));
    request.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
    request.on("error", reject);
  });
}

const server = http.createServer(async (request, response) => {
  if (request.method !== "POST") {
    response.writeHead(405, { "Content-Type": "text/plain" });
    response.end("Method not allowed");
    return;
  }

  const body = await readBody(request);
  const signature = request.headers["x-convo-signature"];

  if (!verifySignature(Array.isArray(signature) ? signature[0] : signature, body)) {
    response.writeHead(401, { "Content-Type": "text/plain" });
    response.end("Invalid signature");
    return;
  }

  const payload = JSON.parse(body);
  console.log("Verified Convo webhook:", payload.event);
  response.writeHead(200, { "Content-Type": "text/plain" });
  response.end("OK");
});

server.listen(port, () => {
  console.log(`Listening for Convo webhooks on http://localhost:${port}`);
});

File: docs/examples/webhook-verify/verify.py
#!/usr/bin/env python3
import hashlib
import hmac
import http.server
import json
import os
import re
import socketserver
import sys
import time

SECRET = os.environ.get("CONVO_WEBHOOK_SECRET")
PORT = int(os.environ.get("PORT", "3000"))
TOLERANCE_SECONDS = 300

if not SECRET:
    print("Set CONVO_WEBHOOK_SECRET before starting the server.", file=sys.stderr)
    sys.exit(1)


def parse_signature_header(header):
    if not header:
        return None

    fields = {}
    for part in header.split(","):
        pieces = part.split("=")
        if len(pieces) != 2:
            return None
        fields[pieces[0]] = pieces[1]

    timestamp_raw = fields.get("t")
    signature = fields.get("v1")
    if not timestamp_raw or not re.fullmatch(r"\d+", timestamp_raw):
        return None
    if not signature or not re.fullmatch(r"[a-f0-9]{64}", signature):
        return None

    return int(timestamp_raw), signature


def sign_body(timestamp, body):
    signed_payload = f"{timestamp}.".encode("utf-8") + body
    return hmac.new(
        SECRET.encode("utf-8"),
        signed_payload,
        hashlib.sha256,
    ).hexdigest()


def verify_signature(header, body):
    parsed = parse_signature_header(header)
    if not parsed:
        return False

    timestamp, signature = parsed
    now = int(time.time())
    if abs(now - timestamp) > TOLERANCE_SECONDS:
        return False

    expected = sign_body(timestamp, body)
    return hmac.compare_digest(signature, expected)


class WebhookHandler(http.server.BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers.get("content-length", "0"))
        body = self.rfile.read(content_length)
        signature = self.headers.get("x-convo-signature")

        if not verify_signature(signature, body):
            self.send_response(401)
            self.end_headers()
            self.wfile.write(b"Invalid signature")
            return

        payload = json.loads(body.decode("utf-8"))
        print("Verified Convo webhook:", payload.get("event"))
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b"OK")

    def do_GET(self):
        self.send_response(405)
        self.end_headers()
        self.wfile.write(b"Method not allowed")


class WebhookHTTPServer(http.server.HTTPServer):
    def server_bind(self):
        socketserver.TCPServer.server_bind(self)
        self.server_name = self.server_address[0]
        self.server_port = self.server_address[1]


if __name__ == "__main__":
    server = WebhookHTTPServer(("127.0.0.1", PORT), WebhookHandler)
    print(f"Listening for Convo webhooks on http://localhost:{PORT}")
    server.serve_forever()
